src/Features/Core/Portable/CodeAnalysisSuggestions/CodeAnalysisSuggestionsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,183 @@
+        ImmutableArray<(string, ImmutableDictionary<string, ImmutableArray<DiagnosticData>>)> configData,
+        Document document,
+        ICodeFixService codeFixService,
+        CancellationToken cancellationToken)
+    {
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+        var location = root.GetLocation();
+
+        using var _1 = ArrayBuilder<CodeAction>.GetInstance(out var actionsBuilder);
+        using var _2 = ArrayBuilder<CodeAction>.GetInstance(out var nestedActionsBuilder);
+        using var _3 = ArrayBuilder<CodeAction>.GetInstance(out var nestedNestedActionsBuilder);
+        var totalViolations = 0;
+        foreach (var (category, diagnosticsById) in configData)
+        {
+            var totalViolationsForCategory = 0;
+            foreach (var (id, diagnostics) in diagnosticsById)
+            {
+                Debug.Assert(diagnostics.All(d => string.Equals(d.Category, category, StringComparison.OrdinalIgnoreCase)));
+
+                var (diagnosticData, documentForFix) = GetPreferredDiagnosticAndDocument(diagnostics, document);
+                var diagnostic = await diagnosticData.ToDiagnosticAsync(document.Project, cancellationToken).ConfigureAwait(false);
+                if (SuppressionHelpers.IsNotConfigurableDiagnostic(diagnostic))
+                    continue;
+
+                if (documentForFix != null)
+                {
+                    var codeFixCollection = await codeFixService.GetDocumentFixAllForIdInSpanAsync(documentForFix, diagnostic.Location.SourceSpan, id, diagnostic.Severity, CodeActionOptions.DefaultProvider, cancellationToken).ConfigureAwait(false);
+                    if (codeFixCollection != null)
+                    {
+                        nestedNestedActionsBuilder.AddRange(codeFixCollection.Fixes.Select(f => f.Action));
+                    }
+                    else
+                    {
+                        // This diagnostic has no code fix, or fix application failed.
+                        // TODO: Can we add some code action such that it shows a preview of the diagnostic span with squiggle?
+                    }
+                }
+                else
+                {
+                    // This is either a project diagnostic with no location OR a dummy diagnostic created from descriptor without background analysis.
+                    // Append this document's span as it location so we can show configure severity code fix for it.
+                    diagnostic = Diagnostic.Create(diagnostic.Descriptor, location);
+                }
+
+                var nestedNestedAction = ConfigureSeverityLevelCodeFixProvider.CreateSeverityConfigurationCodeAction(diagnostic, document.Project);
+                nestedNestedActionsBuilder.Add(nestedNestedAction);
+
+                var totalInCurrentProject = diagnostics.Where(dd => dd.ProjectId == document.Project.Id).Count();
```

Comment:
how would we get diasgnositcs for another project?  that worries me.  how much computation is happening here?

Reply:
Cleaned up now, with appropriate asserts added.